sdk/core/azure-core/tests/async_tests/test_paging_async.py

```diff
@@ -105,39 +267,189 @@ async def extract_data(response):
+            validate_next_request = _validate_token_paging_method,  # arg added for testing purposes
+        )
+
+        assert ['value1.0', 'value1.1', 'value2.0', 'value2.1'] == await _as_list(item_paged)
+
+
+    @pytest.mark.asyncio
+    async def test_next_link_and_continuation_token(self, client,deserialize_output, page_iterator, http_request):
+        def deserialize_output(pipeline_response):
+            if not pipeline_response.http_request.headers.get("x-ms-header", None):
+                return ProductResult(next_link="headerToken,nextLink", value=['value1.0', 'value1.1'])
+            return ProductResult(next_link=None, value=['value2.0', 'value2.1'])
+
+        class ContinueWithRequestHeaderAndNextLink(ContinueWithNextLink):
+            def __init__(self, header_name):
+                super(ContinueWithRequestHeaderAndNextLink, self).__init__()
+                self._header_name = header_name
+
+            def get_next_request(self, continuation_token, initial_request, client):
+                split_token = continuation_token.split(",")
+                token_to_pass_to_headers = split_token[0]
+                next_link = split_token[1]
+                request = super(ContinueWithRequestHeaderAndNextLink, self).get_next_request(next_link, initial_request, client=client)
+                request.headers[self._header_name] = split_token[0]
+                return request
+
+        def _validate_next_link_and_header_paging_method(request):
+            assert request.headers['x-ms-header'] == 'headerToken'
+            assert request.url == "https://baseurl/nextLink"
+
+        item_paged = AsyncItemPaged(
+            deserialize_output=deserialize_output,
+            client=client,
+            page_iterator_class=page_iterator,  # have to add this arg since I'm overriding PageIterator (vast majority won't use this param)
+            initial_state=_get_custom_pipeline_response(headers={'x-ms-token': 'responseToken'}),
+            paging_method=ContinueWithRequestHeaderAndNextLink(header_name='x-ms-header'),
+            continuation_token_location="next_link",
+            validate_next_request=_validate_next_link_and_header_paging_method,  # arg added for testing purposes
+        )
+
+        assert ['value1.0', 'value1.1', 'value2.0', 'value2.1'] == await _as_list(item_paged)
+
+    @pytest.mark.asyncio
+    async def test_cls(self, client, deserialize_output, page_iterator, http_request):
+        def cls(list_of_obj):
+
+            return ["changedByCls"] * len(list_of_obj)
+
+        item_paged = AsyncItemPaged(
+            deserialize_output=deserialize_output,
+            client=client,
+            page_iterator_class=page_iterator,  # have to add this arg since I'm overriding PageIterator (vast majority won't use this param)
+            initial_state=http_request,
+            paging_method=ContinueWithNextLink(),
+            continuation_token_location="next_link",
+            _cls=cls,
```

Comment:
this is def showing the need to find a way around the `cls` issues. going to look into this more tomorrow